### PR TITLE
#34  add expenses group route

### PR DESCRIPTION
**Pull request by galdd**
I add expenses-list route and model

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -32,6 +33,7 @@
     "express": "^4.18.2",
     "express-oauth2-jwt-bearer": "^1.6.0",
     "express-yields": "^1.1.2",
+    "fs": "^0.0.1-security",
```
Comment (ofir-zeitoun) [inline package.json:36]: what is it for?

File: scripts/init-mongo.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+} from "../src/features/expenses-list/expenses-list.model";
+import {
+  ExpensesModel,
+  Expense,
+} from "../src/features/expenses/expenses.model";
+import { UserModel } from "../src/features/users/user.model";
+
+const importData = async () => {
+  try {
+    await mongoose.connect("mongodb://localhost:27017/expanses-01");
```
Comment (ofir-zeitoun) [inline scripts/init-mongo.ts:16]: please use .env so we can work on-prem / cloud

File: scripts/init-mongo.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    const jsonPath = path.join(__dirname, "mock", "expensesListsMock.json");
+    console.log(jsonPath);
+
+    const data = JSON.parse(await fs.readFile(jsonPath, "utf-8"));
+
+    await UserModel.deleteMany();
+    await ExpensesModel.deleteMany();
+    await ExpensesListModel.deleteMany();
+
+    for (const list of data.expensesList) {
```
Comment (ofir-zeitoun) [inline scripts/init-mongo.ts:27]: use map / forEach

File: scripts/init-mongo.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+      if (!isValidObjectId((list.creator as any)._id)) {
```
Comment (ofir-zeitoun) [inline scripts/init-mongo.ts:28]: use filter

File: scripts/init-mongo.ts
Comment (ofir-zeitoun) [inline scripts/init-mongo.ts:1]: too many *any* is used

File: scripts/init-mongo.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  ExpensesModel,
+  Expense,
+} from "../src/features/expenses/expenses.model";
+import { UserModel } from "../src/features/users/user.model";
+
+const importData = async () => {
+  try {
+    await mongoose.connect("mongodb://localhost:27017/expanses-01");
+
+    const jsonPath = path.join(__dirname, "mock", "expensesListsMock.json");
```
Comment (ofir-zeitoun) [inline scripts/init-mongo.ts:18]: can you use import?

File: scripts/init-mongo.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        continue;
+      }
+
+      let creator = await UserModel.findById((list.creator as any)._id);
+      if (!creator) {
+        creator = await UserModel.create(list.creator as any);
+      }
+
+      const expenses = await Promise.all(
+        list.expenses.map(async (expense: { creator: any; _id: any }) => {
```
Comment (ofir-zeitoun) [inline scripts/init-mongo.ts:41]: like on map
any is not valid

File: src/features/expenses-list/expenses-list.model.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
+import mongoose, { Schema, Document } from "mongoose";
+import { Timestamp } from "../../db";
+
+const expensesListSchema = new Schema(
+  {
+    name: { type: String, required: true },
+    creator: { type: Schema.Types.ObjectId, ref: "User", required: true },
+    createdAt: { type: Date, default: Date.now, required: true },
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.model.ts:8]: you can reuse it

File: src/features/expenses-list/expenses-list.model.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    updatedAt: { type: Date, default: Date.now, required: true },
+    expenses: [{ type: Schema.Types.ObjectId, ref: "Expenses" }],
+    users_ids: [{ type: Schema.Types.ObjectId, ref: "User" }],
+  },
+  { versionKey: false, timestamps: true }
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.model.ts:13]: reuse

File: src/features/expenses-list/expenses-list.model.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+);
+
+export type ExpensesList = {
+  name: string;
+  creator?: mongoose.Types.ObjectId;
+  createdAt?: Date;
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.model.ts:19]: reuse, there is a type for it

File: src/features/expenses-list/expenses-list.routes-mock.ts [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+  try {
+    const data = await fs.readFile(jsonPath, "utf8");
+    let expensesListData = JSON.parse(data);
+
+    expensesListData.expenseLists = expensesListData.expenseLists.map(
+      (list: { expenses: any[] }) => {
+        // console.log(list.expenses);
+        const expenseTotal = list.expenses.reduce(
+          (sum: any, { price }: any) => sum + price,
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.routes-mock.ts:23]: use number

File: src/features/expenses-list/expenses-list.routes-mock.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
+import { Router } from "express";
+import { promises as fs } from "fs";
+import path from "path";
+export const router = Router();
+
+router.get("/", async (req, res) => {
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.routes-mock.ts:6]: use zod's valudate

File: src/features/expenses-list/expenses-list.routes.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+} from "./expenses-list.routes-schema";
+import { ExpensesModel } from "../expenses/expenses.model";
+
+export const router = Router();
+
+router.get("/", async (req, res) => {
+  try {
+    console.log("Received request for fetching expense lists.");
+
+    const offset = parseInt(req.query.offset as string, 10) || 0;
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.routes.ts:18]: use zod's validate + define the correct type

File: src/features/stats/stats.routes.ts [hunk 13]
```diff
@@ -0,0 +1,92 @@
+import { Router } from "express";
+import { promises as fs } from "fs";
+import path from "path";
+
+export interface Expense {
```
Comment (ofir-zeitoun) [inline src/features/stats/stats.routes.ts:5]: isn't it duplicated?

**Review by ofir-zeitoun — CHANGES_REQUESTED** [review]
There are some fixes to be done

File: package.json [hunk 14]
```diff
@@ -2,12 +2,13 @@
   "name": "expenses-server",
   "version": "1.0.0",
   "description": "Expanses application's server",
-  "main": "./dist/index.js",
+  "main": "./src/index.ts",
```
[the remark this answers is not in the document]
Comment (galdd) [reply package.json:5]: done

File: package.json (hunk 1 above)
Comment (galdd) [reply package.json:36]: the line removed

File: scripts/init-mongo.ts (hunk 3 above)
Comment (galdd) [reply scripts/init-mongo.ts:27]: done

File: scripts/init-mongo.ts (hunk 2 above)
Comment (galdd) [reply scripts/init-mongo.ts:16]: done

File: scripts/init-mongo.ts (hunk 4 above)
Comment (galdd) [reply scripts/init-mongo.ts:28]: done

File: scripts/init-mongo.ts
Comment (galdd) [reply scripts/init-mongo.ts:1]: fixed

File: scripts/init-mongo.ts (hunk 5 above)
Comment (galdd) [reply scripts/init-mongo.ts:18]: done

File: src/features/expenses-list/expenses-list.model.ts (hunk 7 above)
Comment (galdd) [reply src/features/expenses-list/expenses-list.model.ts:8]: done

File: src/features/expenses-list/expenses-list.routes-mock.ts (hunk 10 above)
Comment (galdd) [reply src/features/expenses-list/expenses-list.routes-mock.ts:23]: the file not necessery anymore, maybe will add again in the test branch

File: src/features/expenses-list/expenses-list.routes-mock.ts (hunk 11 above)
Comment (galdd) [reply src/features/expenses-list/expenses-list.routes-mock.ts:6]: the file not necessery anymore, maybe will add again in the test branch

File: src/features/stats/stats.routes.ts (hunk 13 above)
Comment (galdd) [reply src/features/stats/stats.routes.ts:5]: fixed

File: src/features/expenses-list/expenses-list.model.ts (hunk 9 above)
Comment (galdd) [reply src/features/expenses-list/expenses-list.model.ts:19]: done

File: src/features/expenses-list/expenses-list.model.ts (hunk 8 above)
Comment (galdd) [reply src/features/expenses-list/expenses-list.model.ts:13]: done

File: src/features/expenses/expenses.model.ts [hunk 15]
```diff
@@ -1,26 +1,33 @@
...
-    updatedAt: Date,
+    name: { type: String, required: true },
+    price: { type: Number, required: true },
+    cause: { type: String, required: false },
+    date: { type: Date, required: false },
+    ...dateFields,
+    creator: {
+      type: mongoose.Schema.Types.ObjectId,
+      ref: "User",
+      required: false,
```
Comment (galdd) [inline src/features/expenses/expenses.model.ts:27]: change to true

File: src/features/expenses/expenses.model.ts [hunk 16]
```diff
@@ -1,26 +1,33 @@
...
+    },
   },
   { versionKey: false, timestamps: true }
```
Comment (galdd) [inline src/features/expenses/expenses.model.ts:30]: reuse

File: src/features/expenses/expenses.model.ts [hunk 17]
```diff
@@ -1,26 +1,33 @@
...
 );
 
-export const ExpansesModel = mongoose.model<Expense>(
-  "expenses",
-  expensesSchema
-);
+export const ExpensesModel = mongoose.model<Expense>("Expenses", expenseSchema);
```
Comment (galdd) [inline src/features/expenses/expenses.model.ts:33]: decide the convention

File: src/features/expenses/expenses.model.ts (hunk 15 above)
Comment (galdd) [reply src/features/expenses/expenses.model.ts:27]: done

File: src/features/expenses/expenses.model.ts (hunk 16 above)
Comment (galdd) [reply src/features/expenses/expenses.model.ts:30]: done

File: src/features/expenses/expenses.model.ts (hunk 17 above)
Comment (galdd) [reply src/features/expenses/expenses.model.ts:33]: by the documantion of mongoose its common to use CamelCase
so I left it as it is

File: src/features/expenses-list/expenses-list.routes-schema.ts [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    }),
+  creator: z.object({
+    _id: z.string(),
+    name: z.string(),
+    image: z.string(),
+  }),
+  expenses: z.array(ExpenseSchema),
+});
+
+export const querySchema = z.object({
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.routes-schema.ts:76]: paginationSchema

File: src/features/expenses-list/expenses-list.routes-schema.ts [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+});
+
+export const expensesListIdSchema = z.object({
+  id: z.string().regex(/^[0-9a-fA-F]{24}$/, "ID must be a valid ObjectID"),
+});
+
+export const updateExpensesListSchema =
+  baseExpensesListSchemaNoId.merge(expensesListIdSchema);
+
+export const ExpenseSchema = z.object({
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.routes-schema.ts:30]: needed?

File: src/features/expenses-list/expenses-list.routes-schema.ts [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+
+export const querySchema = z.object({
+  query: z.object({
+    offset: z.coerce.number().optional(),
+    limit: z.coerce.number().optional(),
+    sortOrder: z.enum(["asc", "desc"]).optional(),
+  }),
+});
+
+export type JSONExpense = z.infer<typeof ExpenseSchema>;
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.routes-schema.ts:84]: needed?

File: src/features/expenses-list/expenses-list.routes-schema.ts [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+export type JSONExpensesList = z.infer<typeof ExpensesListSchema>;
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.routes-schema.ts:85]: rename

File: src/features/expenses-list/expenses-list.routes-schema.ts [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+});
+
+export const querySchema = z.object({
+  query: z.object({
+    offset: z.coerce.number().optional(),
+    limit: z.coerce.number().optional(),
+    sortOrder: z.enum(["asc", "desc"]).optional(),
+  }),
+});
+
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.routes-schema.ts:83]: add here:
```
interface QueryParams extends qs.ParsedQs {
  offset: string;
  limit: string;
  sortOrder: "asc" | "desc";
}
```

File: src/features/expenses-list/expenses-list.routes.ts [hunk 23]
```diff
@@ -0,0 +1,65 @@
+import { Router, Request, Response } from "express";
+import status from "http-status";
+import { validateResource } from "../../routes/middlewares";
+import { querySchema } from "./expenses-list.routes-schema";
+import { Expense } from "../expenses/expenses.model";
+import { ExpensesListModel } from "./expenses-list.model";
+
+export const router = Router();
+interface QueryParams extends qs.ParsedQs {
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.routes.ts:9]: move to route file

File: src/features/expenses-list/expenses-list.routes-schema.ts [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+  name: z.string(),
+  price: z.number(),
+  creator: z.object({
+    _id: z.string(),
+    name: z.string(),
+    image: z.string(),
+  }),
+});
+
+export const ExpensesListSchema = z.object({
```
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.routes-schema.ts:53]: camelCase

File: src/features/expenses-list/expenses-list.routes-schema.ts (hunk 19 above)
Comment (ofir-zeitoun) [inline src/features/expenses-list/expenses-list.routes-schema.ts:30]: camelCase

File: src/features/expenses-list/expenses-list.routes-schema.ts (hunk 21 above)
Comment (galdd) [reply src/features/expenses-list/expenses-list.routes-schema.ts:85]: removed

**Review by ofir-zeitoun — APPROVED** [review]

**Pull request merged**